src/TransitionSystems/transition_system.rs

```diff
@@ -1,256 +1,344 @@
+            }
+            LocationID::Quotient(left, right) => {
+                match *(*left) {
+                    LocationID::Simple(_) => write!(f, "{}", (*left))?,
+                    _ => write!(f, "({})", (*left))?,
+                };
+                write!(f, "\\\\")?;
+                match *(*right) {
+                    LocationID::Simple(_) => write!(f, "{}", (*right))?,
+                    _ => write!(f, "({})", (*right))?,
+                };
+            }
+            LocationID::Simple(name) => write!(f, "{}", name)?,
+        }
+        Ok(())
     }
+}
+
+#[derive(Debug, Clone, std::cmp::PartialEq, Eq, Hash, Copy)]
+pub enum CompositionType {
+    Conjunction,
+    Composition,
+    Quotient,
+}
 
-    pub fn set_location(&mut self, index: usize, new_loc: &'a Location) {
-        self.locations[index] = new_loc;
+#[derive(Debug, Clone)]
+pub struct LocationTuple {
+    pub id: LocationID,
+    invariant: Option<Federation>,
+    pub loc_type: LocationType,
+    left: Option<Box<LocationTuple>>,
+    right: Option<Box<LocationTuple>>,
+}
+
+impl PartialEq for LocationTuple {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id && self.loc_type == other.loc_type
     }
+}
 
-    pub fn simple(location: &'a Location, declaration: &Declarations) -> Self {
+impl LocationTuple {
+    pub fn simple(location: &Location, decls: &Declarations, dim: u32) -> Self {
+        let invariant = if let Some(inv) = location.get_invariant() {
+            //println!("invariant of loc {}: {}", location.get_id(), inv);
+            let mut fed = Federation::full(dim);
+            apply_constraints_to_state(&inv, decls, &mut fed);
+            Some(fed)
+        } else {
+            None
+        };
         LocationTuple {
-            locations: vec![location],
-            declarations: vec![declaration.clone()],
+            id: LocationID::Simple(location.get_id().clone()),
+            invariant,
+            loc_type: location.get_location_type().clone(),
+            left: None,
+            right: None,
         }
     }
 
-    pub fn compose(left: Self, right: Self) -> Self {
-        let mut locations = left.locations;
-        locations.extend(right.locations);
-        let mut declarations = left.declarations;
-        declarations.extend(right.declarations);
+    //Merge two locations keeping the invariants seperate
+    pub fn merge(left: &Self, right: &Self, comp: CompositionType) -> Self {
+        /*println!(
+            "Merging {} inv {} and {} inv {}",
+            left.id,
+            left.get_invariants().unwrap_or(&Federation::full(1)),
+            right.id,
+            right.get_invariants().unwrap_or(&Federation::full(1))
+        );*/
+        let id = match comp {
+            CompositionType::Quotient => {
+                LocationID::Quotient(Box::new(left.id.clone()), Box::new(right.id.clone()))
+            }
+            _ => panic!("Invalid merge type {:?}", comp),
```

Comment:
If this function only accepts a quotient CompositionType i think the `comp` parameter should be removed and function name updated to reflect the CompositionType e.g. `merge_quotient` or `merge_into_quotient`

Reply:
Done